File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque
def solution(s):
    n = len(s)
    q = deque(s)
    answer = 0
    
    def test(que):
        dic = {")" : "(", "}" : "{", "]" : "["}
        cur_q = deque()
        for x in que:
            if not cur_q:
                cur_q.append(x)
            else:
                if x in dic and dic[x] == cur_q[-1]:
                    cur_q.pop()
                else:
                    cur_q.append(x)
        if not cur_q:
            return 1
        else:
            return 0
        
            
    for i in range(n):
        q.rotate(-1)
        que = list(q)
        answer += test(que)
    
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;
class Solution {
    public int solution(String s) {
        int answer = 0;
        String ss = s + s;
        Map<Character,Character> map = new HashMap<>();
        map.put(')','(');
        map.put('}','{');
        map.put(']','[');
        for(int i=0; i<s.length(); i++){
            Deque<Character> stack = new ArrayDeque<>();
            for(int j=i; j<i+s.length(); j++){
                char c = ss.charAt(j);
                if(!stack.isEmpty() && map.containsKey(c) && stack.peek() == map.get(c)){
                    stack.pop();
                } else {
                    stack.push(c);
                }
            }
            if(stack.isEmpty()){
                answer++;
            }
        }
        return answer;
    }
}